src/ast/mod.rs

```diff
@@ -3419,31 +3419,67 @@ impl fmt::Display for CreateFunctionUsing {
 /// [1]: https://jakewheat.github.io/sql-overview/sql-2016-foundation-grammar.html#schema-definition
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub enum SchemaName {
+pub enum SchemaNameClause {
     /// Only schema name specified: `<schema name>`.
-    Simple(ObjectName),
+    Simple(SchemaName),
```

Comment:
@alamb to be honest, no.

But one thing that bothers me is that we are ambiguous about structures that we do know the state, but we pass to the API as an ambiguous value. I have two problems with the object name:
- Why it is a tuple? It doesn't make sense to me, and accessing it is a pain. The only gain is the more generic name, but I don't think that's really a huge one.
- The internal vector can be empty, that's my hugest problem. We parse expressions that are REQUIRED to have a name there (otherwise we use options) but, anyway, we have a vector that can be empty. This way the API needs to validate the same thing we do.

I'm not trying to remove support or anything like that. If you tell me that there's a sql that support 3 qualifications besides the table name, we can expand it. But I never saw things like:

`name_1.name_2.name_3.name_4.database.schema.table`

As this is the SCHEMA name, I'm restricting this to `database.schema.table`. This is the fully qualified name for most databases, AFAIK.

Again, my main problem is the 2 points I stated above. The additional things like the differentiation between states and limitation of qualification are additional things that I really wanted, but aren't restricted to be exactly like that.

Reply:
It sounds like perhaps we should change `ObjectName` to be more of an enum style 🤔

Reply:
@alamb there are two problems with this approach:
- Changing the object name means changing ALL statements. I prefer to do it gradually, so the upstream can change it in paces. Changing object name is a huge refactor, while changing it statement by statement is smaller and way more feasible.
- Currently we use object name for everything. How would we differentiate between schema names (that can, mostly, have 1 qualification, 2 tops), table names (that can have 2 or 3 qualifications), and even other structures? We use it for everything which works but is TOO generic without the need of it.

Reply:
🤔

Reply:
@alamb first of all. Replacing the object name is really feasible? I mean, doing statement-by-statement seems to be more reasonable and API-friendly.

Just to clarify, I want to make different structures for `SchemaName`, `TableName`, and anything that can have its own syntax. And, mainly, stop using vectors in places we KNOW there's something.

Reply:
From my perspective, making changes statement-by-statement will simply prolong the downstream churn, but the total amount of churn would be the same

In general I have tried to keep required downstream changes minimal unless required -- typically when people add support for new features that require changing the structs downstream changes are inevitable.

I actually really worry about making sqlparser *less* permissive for names because  I worry that we will
1. Annoy existing users who have to change their code 
2. Really annoy existing users who may be relying on the old behavior of parsing names with three periods for schemas (not sure if there are any people)

Reply:
@alamb can't we find a way to store things that are certainly there, but with variable size?

Also, as I mentioned, if you think there's someone that uses more qualifications, maybe we could have something like:

```rust
pub enum SchemaName {
  /// <schema>
  Simple(Ident),
  /// <database>.<schema>
  DatabaseQualified(Ident, Ident),
  /// [[name_1.[name_2.[name_3.]]]...<name_n>.<database>.<schema>
  CustomQualified(Vec<Ident>, Ident, Ident, Ident)
}
```

This way we can guarantee simple access to the names where we know they are (e.g., `Simple` CERTAINLY have a schema name), while still allowing the upstream to build names as they please.

Reply:
@AugustoFKL  yes I think that would be a resonable change (though definitely a larger downstream change)

Reply:
Ok! For object name I think it could be something like

```rust
pub enum ObjectName {
  Unqualified(Ident),
  Qualified(SchemaName, Ident),
}
```

or 

```rust
pub struct ObjectName {
  name: Ident,
  qualification: Option<SchemaName>,
}
```

What do you think?

For now I'll consider that all objects name have the same support for qualification, although I think we may need to re-check some in the future.